assets/wix/Product.wxs

```diff
@@ -62,6 +68,172 @@
+    or `InstallExecuteSequence` if the UI is skipped):
+
+      1. Load the value that was used in the previous installation into a separate property.
+
+      2. `..._SetFromSaved`: If the property does not have a value (meaning it
+         was not specified on the command line) and there is a value from the
+         previous installation, use the value from the previous installation.
+
+      3. `..._SetDefault`: If the property does not have a value (meaning it was not specified 
+         on the command line _and_ there wasn't a value from a previous installation), and
+         the property has a default value, then set the property to the default value.
+
+      4. `..._DeleteIfOff`: For "boolean" properties, if the property value is not `1`, then delete
+         the property. This needs to be done so that the checkbox in the UI is not initially checked. 
+         If the property exists, the checkbox will be initially checked regardless of the property value.
+
+    Then in the `InstallExecuteSequence`:
+
+      1. `..._SetToOffIfMissing`: For boolean properties, if the property does not exist, it is set
+         to `0`. This ensures that the property value that is saved to the registry is not blank.
+         If blank values are saved to the registry, then the `..._SetFromSaved` step won't run because
+         a blank property value is indistinguishable from a property value that does not exist.
+
+    When using "boolean" properties, always check if the property value is equal `1`, instead 
+    of just checking if the property exists. For example, use `FOO=1` instead of just `FOO`.
+    -->
+
+    <!-- Property: INSTALLFOLDER -->
+    <Property Id="PREVIOUS_INSTALLFOLDER">
+      <RegistrySearch Id="GetSavedInstallFolder" Root="HKLM" Key="$(var.InstallerPropertiesKey)" Name="InstallFolder" Type="raw" />
+    </Property>
+    <SetProperty Id="INSTALLFOLDER" Action="InstallFolder_SetFromSaved" After="AppSearch" Value="[PREVIOUS_INSTALLFOLDER]" Sequence="first">
+      not INSTALLFOLDER and PREVIOUS_INSTALLFOLDER
+    </SetProperty>
+
+    <!-- Property: ADD_PATH -->
+    <Property Id="PREVIOUS_ADD_PATH">
+      <RegistrySearch Id="GetSavedAddToPath" Root="HKLM" Key="$(var.InstallerPropertiesKey)" Name="AddToPath" Type="raw" />
+    </Property>
+    <SetProperty Id="ADD_PATH" Action="AddToPath_SetFromSaved" After="AppSearch" Value="[PREVIOUS_ADD_PATH]" Sequence="first">
+      not ADD_PATH and PREVIOUS_ADD_PATH
+    </SetProperty>
+    <SetProperty Id="ADD_PATH" Action="AddToPath_SetDefault" After="AddToPath_SetFromSaved" Value="1" Sequence="first">
+      not ADD_PATH
+    </SetProperty>
+    <SetProperty Id="ADD_PATH" Action="AddToPath_DeleteIfOff" After="AddToPath_SetDefault" Value="{}" Sequence="first">
+      ADD_PATH&lt;&gt;1
+    </SetProperty>
+    <SetProperty Id="ADD_PATH" Action="AddToPath_SetToOffIfMissing" Before="InstallInitialize" Value="0" Sequence="execute">
+      not ADD_PATH
+    </SetProperty>
```

Comment:
This seems unnecessarily complicated. If the property is not defined to any value, it's already unset by Windows Installer's definition, thus not needing all the explicit `...=1` conditions below.

It's also fine to just bind the actual property to the `RegistrySearch`. See https://robmensching.com/blog/posts/2010/5/2/the-wix-toolsets-remember-property-pattern/.

Reply:
The reason for this complexity is explained in the comments above. What that "Remember Property" pattern doesn't handle is allowing the option to be changed via a checkbox in the UI. If a property has a value, the checkbox that uses that property will be initially checked. It doesn't matter whether the property's value is `0` or `1` or something else; if the property exists, the checkbox is checked. So if the option was disabled, we would remember the property value as `0`. On the next install, that property would be initialized to the value of `0`, but because the property exists, the checkbox would initially be checked.

I'm definitely not a WiX expert, so if there's a simpler way to do this, I'd love to know.

Reply:
If 0 is checking the check box, it could be that the [`Integer`](https://learn.microsoft.com/windows/win32/msi/integer-control-attribute) attribute needs to be set on it. It should be able to bind directly if treated as an integer instead of the default string (which is set if it has any value since 0 is actually "0" in that case).

Reply:
😞  Unfortunately that didn't help. I tried it out on the checkbox for `ADD_FILE_CONTEXT_MENU_RUNPOWERSHELL` like so:

```xml
<Control Id="ContextMenuRunPowerShell" Type="CheckBox" X="20" Y="160" Width="290" Height="17" Property="ADD_FILE_CONTEXT_MENU_RUNPOWERSHELL" Integer="yes" CheckBoxValue="1" Text="Add '$(var.PowerShellFileContextSubMenuDialogText)' context menu for PowerShell files"/>
```

And I reduced the property initialization back to just getting the previous value and restoring the previous value:
```xml
<Property Id="PREVIOUS_ADD_FILE_CONTEXT_MENU_RUNPOWERSHELL">
  <RegistrySearch Id="GetSavedAddFileContextMenuRunPowerShell" Root="HKLM" Key="$(var.InstallerPropertiesKey)" Name="AddFileContextMenuRunPowerShell" Type="raw" />
</Property>
<SetProperty Id="ADD_FILE_CONTEXT_MENU_RUNPOWERSHELL" Action="AddFileContextMenuRunPowerShell_SetFromSaved" After="AppSearch" Value="[PREVIOUS_ADD_FILE_CONTEXT_MENU_RUNPOWERSHELL]" Sequence="first">
  PREVIOUS_ADD_FILE_CONTEXT_MENU_RUNPOWERSHELL
</SetProperty>
```

Running the installer like this...
```
msiexec /i PowerShell-7.4.2-test-win-x64.msi ADD_FILE_CONTEXT_MENU_RUNPOWERSHELL=0
```

... causes the checkbox to be initially checked. So does:
* Running with `ADD_FILE_CONTEXT_MENU_RUNPOWERSHELL=1`.
* Having the registry value `AddFileContextMenuRunPowerShell` equal to `0`.
* Having the registry value `AddFileContextMenuRunPowerShell` equal to `1`.

Reply:
That's unfortunate. I'm reaching out to a colleague who has done more Windows Installer UI, but it could be that the extra CAs are needed. It'd be nice to avoid, though. I'll flag this for follow up.

Reply:
MSI, both in general and specifically for check boxes, uses the logic that a null string is false and anything else is true.

> The selected state of the box sets the property either to the value specified in the Value column of the CheckBox table or to the initial value of the property specified in the Property table.

So this behavior seems consistent.

Reply:
I wonder how the [`Integer`](https://learn.microsoft.com/windows/win32/msi/integer-control-attribute) flag is actually supposed to work then.

I'll re-review this, but might play around a bit with a scratch package. The number of CAs necessary to make this work seem like overkill, but maybe not. What you have certainly works.